### PR TITLE
Universally check for hiwire memory leaks in test suite

### PR DESCRIPTION
**Pull request by hoodmane**
I updated `conftest.py` to check that the number of hiwire keys at the end of the test is the same as the number of hiwire keys at the end of the test. This more or less requires that we never leak any hiwire keys. It'll require a fair amount of work to make the test suite pass like this.

### CODE REVIEW

**Comment by hoodmane** [comment]
I think that most of the remaining test failures represent actual memory leaks and not issues with the tests...

**Comment by hoodmane** [comment]
Merging #1334 will probably fix most of the asyncio-related tests, and then there seems to be another buffer memory leak, and some assortment of random other issues.

**Comment by rth** [comment]
This is quite interesting!

I wonder if we should implement some object to take a snapshot the keys in `globals()` and window, and then automatically delete additional keys at the end of the test in the fixture, to avoid having to do that manually. Related to https://github.com/iodide-project/pyodide/issues/703 That way created objects would be automatically cleaned at the end of the test (and you can do these reference counting checks afterwards)

**Comment by hoodmane** [comment]
Yeah it sure does suggest doing that. I think there is a good way to implement #703 now aside from the fact that we would leak the old `globals` dict. I think we can just call `old_globals.clear()` and then we're only leaking an empty dict, which is better than the alternative of forcefully freeing it (if someone *actually* has an old reference lying around, in one case they see in empty dict, in the other case we crash).

**Comment by hoodmane** [comment]
Actually all we probably have to do is call `pyodide.globals.clear()` at the end of every test.

**Comment by hoodmane** [comment]
It would be helpful if the "short test summary info" used some standard sort order for easier comparisons.

**Comment by hoodmane** [comment]
@rth Would appreciate review on this one, I think it's ready to merge. The important stuff is in `conftest.py`, the rest is just a lot of cleanup and minor fixes in the test suite, or already got merged in one of the separate PRs.

File: conftest.py [hunk 1]
```diff
@@ -242,6 +252,42 @@ def get_driver(self):
...
+
+    Pytest issue #5044:
+    https://github.com/pytest-dev/pytest/issues/5044
+    """
+    selenium = None
+    if "selenium" in item._fixtureinfo.argnames:
+        selenium = item.funcargs["selenium"]
+    if "selenium_standalone" in item._fixtureinfo.argnames:
+        selenium = item.funcargs["selenium_standalone"]
+    if selenium and pytest.mark.norefs.mark not in item.own_markers:
```
Comment (rth) [inline conftest.py:273]:
```suggestion
    if selenium and pytest.mark.skip_refcount_check.mark not in item.own_markers:
```
to be more explicit.

File: packages/matplotlib/test_matplotlib.py [hunk 2]
```diff
@@ -1,12 +1,21 @@
+import pytest
+
+
+@pytest.mark.norefs
```
Comment (rth) [inline packages/matplotlib/test_matplotlib.py:4]:
```suggestion
@pytest.mark.skip_refcount_check
```

here and everywhere to be more explicit

File: pytest.ini [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[pytest]
+markers =
+    norefs: Dont run refcount checks
```
Comment (rth) [inline pytest.ini:3]: This should go into setup.cfg which already has a pytest config section

File: src/tests/test_jsproxy.py [hunk 4]
```diff
@@ -389,65 +392,36 @@ def test_mount_object(selenium):
...
+        return result;
         """
     )
     assert result[:3] == ["x1", "x2", 3]
     assert set([x for x in result[3] if len(x) == 1]) == set(["x", "y", "s", "t"])
     assert set([x for x in result[4] if len(x) == 1]) == set(["x", "y", "u", "t"])
-
-
-@pytest.mark.xfail
-def test_mount_map(selenium):
```
Comment (rth) [inline src/tests/test_jsproxy.py:409]: It's redundant with existing tests?

File: src/tests/test_jsproxy.py [hunk 5]
```diff
@@ -739,16 +723,13 @@ def raises(exc, match=None):
 
 
 def test_memory_leaks(selenium):
-    assert 1 == selenium.run_js(
+    selenium.run_js(
```
Comment (rth) [inline src/tests/test_jsproxy.py:726]: Maybe add a comment that refcounts are checked elsewhere

**Review by rth — COMMENTED** [review]
Thanks! A few comments, otherwise LGTM.

File: src/tests/test_jsproxy.py (hunk 4 above)
Comment (hoodmane) [reply src/tests/test_jsproxy.py:409]: It tests for behavior that was never implemented and that I'm not really planning to implement. Note the `xfail` marker.

**Pull request merged**